### ISSUE TITLE
Add carthage build argument to override "SWIFT_TREAT_WARNINGS_AS_ERRORS" 

### ISSUE THREAD

**Issue by 1oo7** [Carthage/Carthage#2922]
* carthage install method: [ ] .pkg, [x] homebrew, [ ] source
* `which carthage`: /usr/local/bin/carthage
* `carthage version`: 0.34.0
* `xcodebuild -version`: Xcode 11.3 Build version 11C29
* Are you using `--no-build`? NO
* Are you using `--no-use-binaries`? YES
* Are you using `--use-submodules`? NO
* Are you using `--cache-builds`? NO
* Are you using `--new-resolver`? NO

**Cartfile**
```
github "ReactiveCocoa/ReactiveSwift" == 5.0.1
```

**Carthage Output**
```
Build Failed
	Task failed with exit code 65:
	/usr/bin/xcrun xcodebuild -workspace /Users/mac-daddie/Develop/ios/Carthage/Checkouts/ReactiveSwift/ReactiveSwift.xcworkspace -scheme ReactiveSwift-watchOS -configuration Release -derivedDataPath /Users/mac-daddie/Library/Caches/org.carthage.CarthageKit/DerivedData/11.3_11C29/ReactiveSwift/5.0.1 -sdk watchos ONLY_ACTIVE_ARCH=NO CODE_SIGNING_REQUIRED=NO CODE_SIGN_IDENTITY= CARTHAGE=YES archive -archivePath /var/folders/w7/d9dqqx_938qfd43_vrf060900000gp/T/ReactiveSwift SKIP_INSTALL=YES GCC_INSTRUMENT_PROGRAM_FLOW_ARCS=NO CLANG_ENABLE_CODE_COVERAGE=NO STRIP_INSTALLED_PRODUCT=NO (launched in /Users/mac-daddie/Develop/ios/Carthage/Checkouts/ReactiveSwift)

This usually indicates that project itself failed to compile. Please check the xcodebuild log for more details: /var/folders/w7/d9dqqx_938qfd43_vrf060900000gp/T/carthage-xcodebuild.mzgl8j.log
```

In the log:
```
/Users/mac-daddie/Develop/ios/Carthage/Checkouts/ReactiveSwift/Sources/Signal.swift:1852:71: error: will never be executed
                                lifetime += action(index, strategy) { observer.send($0.map { _ in fatalError() }) }
                                                                                                  ^
/Users/mac-daddie/Develop/ios/Carthage/Checkouts/ReactiveSwift/Sources/Signal.swift:1852:64: note: '_' is uninhabited, so this function body can never be executed
                                lifetime += action(index, strategy) { observer.send($0.map { _ in fatalError() }) }
```

**Actual outcome**
Carthage did not build ReactiveSwift 5.0.2 because ReactiveSwift's xcconfig has `SWIFT_TREAT_WARNINGS_AS_ERRORS = YES`. 

**Expected outcome**
Carthage should build ReactiveSwift 5.0.2 because it is an officially released version that is required by my project. It should not let minor warnings stop it from building required dependencies. 

Rationale:

We all know Apple loves to add new warnings when they update the Swift compiler. This is nothing new. It's useful for developers, but open source projects should never get released with `SWIFT_TREAT_WARNINGS_AS_ERRORS = YES` because it basically guarantees that framework will break in some future XCode update, over a minor warning.

To keep dependencies being buildable across XCode updates, `SWIFT_TREAT_WARNINGS_AS_ERRORS = NO` needs to be enforced by Carthage. 

In the case of ReactiveSwift, simply deactivating this one build setting makes version 5.0.2 perfectly compatible with XCode 11.3 and Swift 5.1 compiler builds. 

Updating a major XCode version is already annoying enough, without having new warnings break everything because Carthage can't override a build flag that's not for released software.

On the other hand, when you're doing dependency development using submodules, you would want SWIFT_TREAT_WARNINGS_AS_ERRORS not to be overriden to NO. Because in this case, you can actually do something about any warnings and you may want to make sure you don't miss any (which is the point of that setting).

**Comment by gistya** [comment]
A PR has been made to fix this issue: https://github.com/Carthage/Carthage/pull/2924

**Comment by jdhealy** [comment]
You can (all likelihood) enact this yourself on all Carthage-invoked xcodebuilds via `XCODE_XCCONFIG_FILE`:

```shell
$ cat >! ~/Library/Preferences/platform-unspecific-swift-treat-warnings-as-errors.xcconfig <<-'EOF'
SWIFT_TREAT_WARNINGS_AS_ERRORS[sdk=*] = NO
EOF

$ XCODE_XCCONFIG_FILE="${HOME}/Library/Preferences/platform-unspecific-swift-treat-warnings-as-errors.xcconfig" carthage …
```

**Comment by jdhealy** [comment]
You can (all likelihood) enact this yourself on all Carthage-invoked xcodebuilds via `XCODE_XCCONFIG_FILE`:

```shell
$ cat >! ~/Library/Preferences/platform-unspecific-override-to-unset·swift-treat-warnings-as-errors.xcconfig <<-'EOF'
SWIFT_TREAT_WARNINGS_AS_ERRORS[sdk=*] = NO
EOF

$ XCODE_XCCONFIG_FILE="${HOME}/Library/Preferences/platform-unspecific-override-to-unset·swift-treat-warnings-as-errors.xcconfig" carthage …
```

**Comment by 1oo7** [comment]
@jdhealy then why doesn't Carthage use that method instead of setting `-suppress-warnings` in its own Makefile? 

The reason is because you don't want someone to have to add global cruft to their Library when what you really want is just for, say, a single dependency to have this setting overridden to ON or OFF. 

My PR for this issue does just that: it allows you to say
`carthage build ReactNative --no-use-binaries --warnings-as-errors no`

So that you can build ReactNative 5 without errors, just like Carthage does when you build Carthage itself. (etc.)

We also want to have this option for the CI pipeline to be able to clear through issues without making a global setting that becomes forgotten and then may cause unexpected concerns later.

**Comment by jdhealy** [comment]
> @jdhealy then why doesn't Carthage use that method instead of setting -suppress-warnings in its own Makefile?

That's because Carthage-the-Tool isn't built using xcodebuild currently (see the Swift Package Manager invocation), so we pass what you mentioned as a SwiftPM flag.

> you don't want someone to have to add global cruft to their Library

Sidenote: you can place the `$XCODE_XCCONFIG_FILE` elsewhere on disk.

> …

Haven't weighed in on the PR yet, just wanted to post a quick comment and let people know there's a workflow available via the current release. Again, haven't weighed in on the PR yet.

**Comment by 1oo7** [comment]
> we pass what you mentioned as a SwiftPM flag

That's what we want to be able to do as well, but via Carthage, since we don't use Swift PM. 

I realized that the original title of this issue was misleading—I've updated it to say **Add carthage build argument to override "SWIFT_TREAT_WARNINGS_AS_ERRORS"*, since it would not be good to always force it to off. 

> Haven't weighed in on the PR yet, just wanted to post a quick comment and let people know there's a workflow available via the current release. Again, haven't weighed in on the PR yet.

Thanks, I understand, and I appreciate your time on the review. I did try to make it a solid PR, and in the spirit of Carthage. 

The workaround I have been using up until now is to `carthage checkout`, then manually set the settings in various projects in Carthage/Checkouts/ via `sed`/`awk`/`php -R preg_replace()` or by hand in XCode, then `carthage build`. 

But we really prefer if, like with SPM, this could be passed through as a flag at build time as needed. That would be much more maintainable and elegant.
